File: content/exercises/intermediate-python/pandas/dataframes/checker.py
import io
import runpy
import sys


def main() -> None:
    submission_path = sys.argv[1]
    output_path = sys.argv[2]
    tests: list[tuple[str, bool, str]] = []
    feedback: list[str] = []
    status = "failed"

    try:
        namespace = runpy.run_path(submission_path, run_name="__main__")
        tests.append(("code runs", True, "Code executed without errors."))
    except Exception as error:  # noqa: BLE001
        namespace = {}
        tests.append(("code runs", False, "Code did not run successfully."))
        feedback.append(f"Execution error: {error}")

    if namespace:
        df = namespace.get("df")
        if df is None:
            tests.append(("df exists", False, "Define a pandas DataFrame named df."))
        else:
            try:
                import numpy as np
                import pandas as pd

                is_df = isinstance(df, pd.DataFrame)
                tests.append(("df type", is_df, "df should be a pandas DataFrame."))
                if is_df:
                    has_col = "gdp_per_capita" in df.columns
                    tests.append(
                        ("gdp_per_capita column", has_col, "Add a gdp_per_capita column to df.")
                    )
                    if has_col:
                        expected = df["gdp_billion"] * 1000 / df["population_million"]
                        values_ok = np.allclose(df["gdp_per_capita"].to_numpy(), expected.to_numpy())
                        tests.append(
                            (
                                "gdp_per_capita values",
                                values_ok,
                                "gdp_per_capita should equal gdp_billion * 1000 / population_million.",
                            )
                        )
            except Exception as error:  # noqa: BLE001
                tests.append(("pandas validation", False, f"Could not validate df: {error}"))

    passed_all = tests and all(item[1] for item in tests)
    if passed_all:
        status = "passed"
        feedback.append("Correct solution submitted.")
    elif not feedback:
        feedback.append("Create df and add gdp_per_capita with column arithmetic.")

    output_stream = io.StringIO()
    output_stream.write(f"STATUS:{status}\n")
    for item in feedback:
        output_stream.write(f"FEEDBACK:{item}\n")
    for name, passed, message in tests:
        output_stream.write(f"TEST:{name}|{'pass' if passed else 'fail'}|{message}\n")

    with open(output_path, "w", encoding="utf-8") as handle:
        handle.write(output_stream.getvalue())


if __name__ == "__main__":
    main()

File: content/exercises/intermediate-r/dplyr/filter-mutate/checker.R
args <- commandArgs(trailingOnly = TRUE)

if (length(args) < 2) {
  stop("Usage: checker.R <submission_path> <output_path>")
}

submission_path <- args[[1]]
output_path <- args[[2]]

test_results <- list()
feedback <- c()
status <- "failed"

record_test <- function(name, passed, message) {
  test_results[[length(test_results) + 1]] <<- list(name = name, passed = passed, message = message)
}

env <- new.env(parent = baseenv())
execution_ok <- TRUE
execution_error <- NULL

tryCatch(
  {
    source(submission_path, local = env)
  },
  error = function(e) {
    execution_ok <<- FALSE
    execution_error <<- conditionMessage(e)
  }
)

if (!execution_ok) {
  record_test("code runs", FALSE, "Your code did not run successfully.")
  feedback <- c(feedback, paste("Execution error:", execution_error))
} else {
  record_test("code runs", TRUE, "Code executed without errors.")
  has_result <- exists("result", envir = env, inherits = FALSE)
  record_test("result exists", has_result, "Define filtered data frame as `result`.")
  if (has_result) {
    result <- get("result", envir = env, inherits = FALSE)
    is_df <- is.data.frame(result)
    has_kpl <- is_df && "kpl" %in% colnames(result)
    mpg_filtered <- is_df && all(result$mpg > 20)
    record_test("result is data frame", is_df, "`result` should be a data frame.")
    record_test("kpl column", has_kpl, "Add kpl = mpg * 0.425.")
    record_test("mpg filter", mpg_filtered, "Filter rows where mpg > 20.")
  }
}

passed_all <- length(test_results) > 0 && all(vapply(test_results, function(t) isTRUE(t$passed), logical(1)))
if (passed_all) {
  status <- "passed"
  feedback <- c(feedback, "Correct solution submitted.")
} else if (length(feedback) == 0) {
  feedback <- c(feedback, "Create `result` using filter(mpg > 20) and mutate(kpl = mpg * 0.425).")
}

output_lines <- c(paste0("STATUS:", status))
for (msg in feedback) output_lines <- c(output_lines, paste0("FEEDBACK:", msg))
for (t in test_results) {
  output_lines <- c(output_lines, paste0("TEST:", t$name, "|", if (isTRUE(t$passed)) "pass" else "fail", "|", t$message))
}
writeLines(output_lines, con = output_path)
